server/routes/auth.ts

```diff
@@ -27,10 +29,42 @@ router.get(
     failureRedirect: process.env.MARKETPLACE_APP_URL,
   }),
   function (req, res) {
+    updateActiveAccounts(req);
     res.redirect(`${process.env.MARKETPLACE_APP_URL}/profile`);
   },
 );
 
+router.get('/accounts', ensureLoggedIn(), (req, res) => {
+  if (!req.session) {
+    return res.sendStatus(500).json({ error: 'req.session is falsy' });
+  }
+  return res.json({
+    activeAccountId: req.session.activeAccountId,
+    authenticatedAccountIds: req.session.authenticatedAccountIds,
```

Comment:
We may want to add more information about each account in the future like the email &/or address, so maybe we should do something like:
```suggestion
      authenticatedAccountIds: req.session.authenticatedAccountIds.map(id => {id:id}),
```

Reply:
@aaronc the design thought that Marie and I had here was that in the cookie, we should only specify an account id/s. the idea being that it's always left up to the client-side to fetch more information about the currently authenticated user. this is an additional fetch request. BUT the major benefit is that our client-side has an autogenerated graphql API with types and all that. the benefit is that we can just use the autogenerated type `account` type after doing an auto-generated `accountById` graphql query.

Reply:
hope that makes sense, basically the idea, is that we didn't want to have to maintain a custom type for the REST API endpoint in regen-web, and instead be able to make use of the graphql generated type

Reply:
Well I think the issue though is that we can't assume that we can fetch info about the other authenticated accounts without actually switching. So the current account ID could fetch it yes, but the other IDs maybe not. So I think this one API endpoint needs to provide some data which is only available with the context that it has

Reply:
currently any app user is able to read accounts info without any further authz since we need that for public profiles, eg https://app.regen.network/profiles/regen1dlszg2sst9r69my4f84l3mj66zxcf3umcgujys30t84srg95dgvs8rn9rj/portfolio

Reply:
We can fetch account info for any id:
```
    POLICY "account_select_all" FOR SELECT
      USING (true)
```

Reply:
but I guess instead the idea would be to only GRANT SELECT on only certain columns that we can have public but limit access to some other columns like email, google id, etc.
we didn't really need this granularity so far with web3 users but it might be worth looking into for web2 users

Reply:
Emails should _definitely_ not be part of the public profile. So I think this needs to be handled in this endpoint. We probably need to split account data into a public profile table and a private profile table. I think that will make more sense than a GRANT SELECT on specific columns although maybe that could work too.

Reply:
> but I guess instead the idea would be to only GRANT SELECT on certain columns that we can have public but limit access to some other columns like email, google id, etc.
we didn't really need this granularity so far with web3 users but it might be worth looking into for web2 users

makes sense, that would be good to handle soon, but probably could be done separately

Reply:
Definitely can be done separately, but doing `{id:id}` makes it easy to add

Reply:
let's assume we just restrict some info, there is still the ability to query all accounts via graphql, you just can't see the restricted fields. i guess if that's a problem then indeed we need to provide more in these responses.

Reply:
well, it's easy to restrict the access to certain columns on the account table with grants for unauthenticated users (via the app_user role).
what's not clear to me is how we would grant access to the private field columns ONLY for the currently authenticated user.

here's an alternative idea, with the goal of being able to utilize autogenerated types AND still satisfying the privacy requirements:

i think we could potentially do this with a custom function and still get auto-generated types.

1. set a postgres session variable equal to an uuid array based on the `authenticatedAccountIds`, i.e.:
```
set authenticatedAccountIds = '{04a91abe-6df8-11ee-8bcd-c26700a3bae4}';
```
(in our case that happens in the postgraphile middleware, easy to do)

2. write a custom function that overrides the grant restrictions on the account table that i mentioned above (the so-called use security definer or whatever), i.e. here's how you load an array of uuids as a valid postgres type:
```
select current_setting('authenticatedAccountIds')::uuid[];
```

the custom function would return the set of accounts where the ids are in the array above.
with that approach, we can still get the automatically generated type for the client-side via graphql.
AND get the private data ONLY in the case of authenticated requests.

we [already have a function](https://github.com/regen-network/regen-server/blob/feat-web2-auth/migrations/schema_snapshot.sql#L368-L372) that gives us the account for the current active account based on the user session:
this new function would enable us to get the account info the authenticated accounts, based on the user session.

i think it could be nice!

Reply:
here's an basic example of what the query could look like:
```
regen_registry=# SELECT  *
FROM    account
WHERE   id = ANY('{04a91abe-6df8-11ee-8bcd-c26700a3bae4,04bda45c-6df8-11ee-8bcd-c26700a3bae4}'::UUID[]);
```
in reality it would look be more like:
```
regen_registry=# SELECT  *
FROM    account
WHERE   id = ANY(current_setting('authenticatedAccountIds')::uuid[]);
```

Reply:
so if I understood correctly, the idea would be to:
- only GRANT SELECT on the public `account` columns to `app_user`, this way we have the level of data privacy we want
- but since the db root user is granted all access, the security definer function you're talking about (that would then map to some graphql query) could return all account info for any authenticated accounts. In this case, we should make sure to follow postgres recommendations in terms of security: https://www.postgresql.org/docs/current/sql-createfunction.html#SQL-CREATEFUNCTION-SECURITY